### ISSUE TITLE
Run fuzz tests on merge to master

### ISSUE THREAD

**Issue by cmichi** [paritytech/ink#535]
## Problem

We now have a couple fuzz tests (which can be run via `cargo test --features ink-fuzz-tests -- fuzz_tests`). They are only executed if tests are run with this feature explicitly enabled though ‒ this was done intentionally to not put load on the CI for every single push on every single PR.

So at the moment there is no guarantee that they are ever executed. We discussed this and came to the conclusion that an ideal case would be:
* They are run once you press "Merge" in the GitHub UI.
* If they are then successful the merge happens, otherwise not.

This way they would only be run for every merge to `master`, and not for every commit like the usual CI tests.

## Solutions
I talked to @TriplEight a while ago about this and he mentioned two apparent solutions:

### Fuzz after merge
Add a CI stage which runs the fuzz tests, but is set to only run on master. This has the drawback of potential errors being only detected after a merge and thus breaking an evergreen master. We would then have to revert the commit manually.

### Bors
Add a bors command, which we would need to post in a GitHub comment. This would then run the fuzz tests and merge automatically if they are fine. This has the drawback of having to post a comment under every PR and discard usage of the big green "Squash & Merge" button.

## IMHO

Both solutions are not ideal, but both are better than the status-quo of not running them at all. I would opt for the "Fuzz after merge" solution, since it doesn't require anybody to think about posting a comment. Special procedures like this always end up being omitted; first for smallish PRs, later for some where you are "kind of sure" that it's not necessary, and eventually it will lose it's value altogether.

**Comment by TriplEight** [comment]
About Bors: 
The ultimate goal of using it is omitting the big green "Squash & Merge" button. That is, merging **everything** by `bors merge` and if you want more tests, `bors try`. That additionally ensures that your branch is tested on a merged with the master branch, and merges if green. Hence, even greener master.
The single downside of falling in Nirvana with Bors is that it is not possible to integrate it into our GitLab "open-pipeline" workflow. This means that either you use Bors, or the external contributors having their CI. It is the reason why Bors is not implemented everywhere. The technical reason lies between three issues: 
- due to GitLab mirroring quirks we can't mirror the branch names and have these pipelines open to the forked PRs;
- due to other GitLab mirroring and branch protection quirks and Bors design we can't mirror the temp branches required by Bors (`trying` and `staging` we can, but `trying.tmp` and `staging.tmp` can't be exported from GitHub AFAIU); 
- due to PRs being mirrored to their numbers and not the branch names, Bors can't receive the pipeline status back

**Comment by TriplEight** [comment]
So the only real way of having "pre-merge" pipeline is to set the GitLab job to `trying` or `staging` branch only. Then make the job obligatory to merge the branch. And running `bors merge` will become mandatory. Many use this approach (not with GitLab tho).

There might be another way instead of running fuzzing on master. Set a schedule job to branch off master and run it there. Additional effort would be that it would be nice from the failed job to file an issue with a link to the failed pipeline.

I'd vote for the latter as a temporary measure and putting the question about Bors to the internal tooling team and `Parity-Processbot`.

**Comment by TriplEight** [comment]
And additional possibility, not very handy though:
- make the fuzzing job always manual, so the PR author will go to GitLab UI and push that play button when needed.
- make this job obligatory for merging.

The former can be automated variously: `parity-processbot` to read some certain github comment (or some certain label) and trigger the manual job, receive the status.

The downside here will be that **all** the not-ready to merge jobs will have the yellow status (waiting for the obligatory job to report status)

**Comment by cmichi** [comment]
>There might be another way instead of running fuzzing on master. Set a schedule job to branch off master and run it there. Additional effort would be that it would be nice from the failed job to file an issue with a link to the failed pipeline.

AFAIU we could also add another fuzz-test job to our GitLab config, which runs only after a merge into the `master` branch, right? Potentially with the automated issue creation you suggested. What advantage does a scheduled job have in contrast to this?

**Comment by TriplEight** [comment]
Yeah, of course it can run every time on master, scheduling is not necessary in this case, would even be confusing since it won't be clear who offended the fuzzing.

**Comment by Robbepop** [comment]
I think the best temporary solution for now is to run the fuzz tests (that take a while to execute) AFTER a merge to `master` branch and report via generated issues upon found bugs.
